cpp/dolfinx/fem/DirichletBC.h

```diff
@@ -495,76 +508,71 @@ class DirichletBC
+      std::span<const T> values = g->x()->array();
+      if (x0.has_value())
       {
-        auto g = std::get<std::shared_ptr<const Function<T, U>>>(_g);
-        assert(g);
-        auto dofs1_g
-            = _dofs1_g.empty() ? std::span(_dofs0) : std::span(_dofs1_g);
-        std::span<const T> values = g->x()->array();
-        if (x0.has_value())
-        {
-          std::span<const T> _x0 = x0.value();
-          assert(x.size() <= _x0.size());
-          for (std::size_t i = 0; i < _dofs0.size(); ++i)
-          {
-            if (_dofs0[i] < x_size)
+        std::span<const T> _x0 = x0.value();
+        assert(x.size() <= _x0.size());
+        apply(
+            [&](std::size_t i) -> T
             {
-              assert(dofs1_g[i] < (std::int32_t)values.size());
-              x[_dofs0[i]] = alpha * (values[dofs1_g[i]] - _x0[_dofs0[i]]);
-            }
-          }
-        }
-        else
-        {
-          for (std::size_t i = 0; i < _dofs0.size(); ++i)
-          {
-            if (_dofs0[i] < x_size)
+              assert(dofs1_g[i] < static_cast<std::int32_t>(values.size()));
+              return alpha * (values[dofs1_g[i]] - _x0[_dofs0[i]]);
+            });
+      }
+      else
+      {
+        apply(
+            [&](std::size_t i) -> T
             {
-              assert(dofs1_g[i] < (std::int32_t)values.size());
-              x[_dofs0[i]] = alpha * values[dofs1_g[i]];
-            }
-          }
-        }
+              assert(dofs1_g[i] < static_cast<std::int32_t>(values.size()));
+              return alpha * values[dofs1_g[i]];
+            });
       }
-      else if (std::holds_alternative<std::shared_ptr<const Constant<T>>>(_g))
+    };
+
+    auto handle_constant = [&](std::shared_ptr<const Constant<T>> g)
+    {
+      const std::vector<T>& value = g->value;
+      std::int32_t bs = _function_space->dofmap()->bs();
+      if (x0.has_value())
       {
-        auto g = std::get<std::shared_ptr<const Constant<T>>>(_g);
-        const std::vector<T>& value = g->value;
-        std::int32_t bs = _function_space->dofmap()->bs();
-        if (x0.has_value())
-        {
-          assert(x.size() <= x0.value().size());
-          std::ranges::for_each(
-              _dofs0,
-              [x_size, &x, x0 = x0.value(), &value, alpha, bs](auto dof)
-              {
-                if (dof < x_size)
-                  x[dof] = alpha * (value[dof % bs] - x0[dof]);
-              });
-        }
-        else
-        {
-          std::ranges::for_each(_dofs0,
-                                [x_size, bs, alpha, &value, &x](auto dof)
-                                {
-                                  if (dof < x_size)
-                                    x[dof] = alpha * value[dof % bs];
-                                });
-        }
+        assert(x.size() <= x0.value().size());
+        apply(
+            [&](std::size_t i) -> T {
+              return alpha * (value[_dofs0[i] % bs] - x0.value()[_dofs0[i]]);
```

Comment:
Should a temporary reference be made to `_dofs0[i]` to avoid duplicate access queries?